### ISSUE TITLE
importing npm module into typescript code silently breaks build

### ISSUE THREAD

**Issue by bamnet** [bazelbuild/rules_nodejs#1959]
# 🐞 bug report

### Affected Rule

 The issue is caused by the rule:

Unknown.

### Is this a regression?

No.

### Description

Adding an module import like `import chart from 'tui-chart';` to my typescript code and an associated `"@npm//tui-chart",` dep to the `ts_library` silently breaks the build.  `ts_devserver` still serves JavaScript but it doesn't eval at all.

I'm 99% sure this is a problem in my setup but `bazel build ...` doesn't show any errors which makes this difficult to debug. There is a note in the Typescript [docs](https://bazelbuild.github.io/rules_nodejs/TypeScript.html#self-managed-npm-dependencies) to see https://github.com/bazelbuild/rules_nodejs#dependencies for more information on managing NPM dependencies but it no longer works.

## 🔬 Minimal Reproduction

https://github.com/bamnet/debug

1. `bazel run :devserver`
2. Visit http://localhost:8080 and observe there is no "Hello World" written to your console.
3. Remove the `import` line at the top of `test.ts`, restart the devserver, and see the console.log works.

## 🔥 Exception or Error

<pre><code>
N/A. Bazel runs without error.

</code></pre>

## 🌍  Your Environment

**Operating System:**

<pre>
  <code>
Ubuntu 18.04.4 LTS
  </code>
</pre>

**Output of `bazel version`:**

<pre>
  <code>
Build label: 3.2.0
Build target: bazel-out/k8-opt/bin/src/main/java/com/google/devtools/build/lib/bazel/BazelServer_deploy.jar
Build time: Wed May 27 17:54:00 2020 (1590602040)
Build timestamp: 1590602040
Build timestamp as int: 1590602040
  </code>
</pre>

**Rules_nodejs version:**

(Please check that you have matching versions between WORKSPACE file and `@bazel/*` npm packages.)

<pre>
  <code>
1.7.0
  </code>
</pre>

**Anything else relevant?**

**Comment by bamnet** [comment]
Feel free to close this and redirect me to StackOverflow. What feels strange to me is that I'm not getting any bazel errors yet there's something wrong with my build.  I struggled to find a very simple example that imports npm modules I could crib off that wasn't the full angular example app (happy to contribute once I figure it out).

**Comment by evahowe** [comment]
ok, thanks @bamnet - I will put this on the list to look into. If you resolve it before then, let us know.

**Comment by bamnet** [comment]
I was able to get this working, but it's not clear if my approach is WAI or just a clever workaround.

My new understanding is that while a ts_library rule needs to depend on the //@npm module, that doesn't result in the module actually being pulled into the build process that produces JS to run in the browser. If you point a simple ts_devserver at your library, it chokes (silently) on the import line. As a result, we need to develop using the rollup bundle.

1. Upgrade to rules_nodejs 2.0.0-rcx because I like to update things. As part of the upgrade, replace references to `@npm_bazel_typescript` with `@npm//bazel/typescript`.
2. Instead of having `ts_devserver` depend on your `ts_library`, use it to serve a static_file from your `rollup__bundle` rule.  ts_library -> rollup_bundle -> ts_devserver (via static_files instead of deps).
3. Tell rollup how to resolve the dependency.  For me, this involved a rollup.config.js file which used `plugins: [commonjs(), nodeResolve()]`.  This might be caused by the older style node module I'm trying to use not exporting things correctly.

I _think_ this detour was largely caused by my use of node modules which don't have cleanly defined modules and exports like angular / vue / etc.

**Comment by alexeagle** [comment]
It's a tradeoff - using rollup in the development server can make the round-trip for making changes slower (though we have an experimental worker mode that makes it fast like rollup's watch mode)
OTOH using `ts_devserver` you have to deal with third-party libraries that can't be expressed as requirejs (AMD) modules. Note that we provide AMD wrappers around all third-party libs (you can bazel query in `@npm//` to see the `npm_umd_bundle rules in there, will be something like `@npm//tui-chart__amd`)

**Comment by applmak** [comment]
I spent a bit of time looking into this on my own project, and afaict, this is the same kind of behavior that was happening to me in karma tests earlier (see #1872). Adding `export={}` to the end of my .ts file fixed it then. I wonder if there is a similar kind of strange thing going on as part of the generation of the amd file which causes the loading of the npm to fail. For reference, the exact behavior in the generated `define`s in my case is:

  1. Try to load main.ts around the code snippet of:
```
if (options.enabled || this.enabled) {
  //Enable this module and dependencies.
  //Will call this.check()
  this.enable();
} else {
  this.check();
}
```
but go to the check case (we are waiting on stuff to load).
  2. Try to load lit-element (same, we are waiting on stuff to load).
  1. Successfully load _@r2 (that is, enable that module)
  1. Successfully load _@r3 (same, enable that module).
  1. Then, nothing.

**Comment by applmak** [comment]
Actually, I think it's that nothing does the final `require('mymainfile')`. My custom index.html looks like this now, and things load just fine:

```
  <script src="index.js"></script>
  <script>
    require(['root/path/to/my/entrypoint'], function(main) {
      console.log('loaded');
    });
  </script>
```

BUILD file excerpt:
```
ts_library(
  name="main",
  srcs=["main.ts"],
  deps = [
    "@npm//lit-element",
  ]
)
ts_devserver(
  name = "devserver",
  # We'll collect all the devmode JS sources from these TypeScript libraries
  deps = [
    ":main",
    "@npm//lit-element:lit-element__umd",
  ],
  # This is the path we'll request from the browser, see index.html
  serving_path = "/index.js",
  # The devserver can serve our static files too
  static_files = ["index.html"],
)
```
